File: 713. Subarray Product Less Than K/numSubarrayProductLessThanK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const numSubarrayProductLessThanK = (nums, k) => {
  if (k <= 1) {
    return 0;
  }

  let count = 0;
  let product = 1;
  let left = 0;

  for (let right = 0; right < nums.length; right++) {
    product *= nums[right];
    while (product >= k) {
      product /= nums[left];
      left++;
    }
    count += right - left + 1;
  }
  
  return count;
};

const nums = [10, 5, 2, 6]
const k = 100
const result = numSubarrayProductLessThanK(nums, k)
console.log('numSubarrayProductLessThanK: ', result)

File: 713. Subarray Product Less Than K/numSubarrayProductLessThanK.py
from typing import List

class Solution:
  def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
    if k <= 1:
      return 0
    
    count = 0
    product = 1
    left = 0
    
    for right in range(len(nums)):
      product *= nums[right]
      while product >= k:
        product /= nums[left]
        left += 1
      count += right - left + 1
      
    return count
      
nums = [10, 5, 2, 6]
k = 100
solution = Solution()
result = solution.numSubarrayProductLessThanK(nums, k)
print('numSubarrayProductLessThanK: ', result)
